### PR TITLE
feat: [COMDEX] add x/wasm module parsing

### PR DESCRIPTION
**Pull request by huichiaotsou**
## Description

Closes: #XXXX

implement `x/wasm` for comdex chain

- Available messages for testing: 
  - MsgStoreCode: height 2979613
  - MsgInstantiateContract: height 3082705
  - MsgExecuteContract: everywhere

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [x] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [x] targeted the correct branch
- [ ] provided a link to the relevant issue or specification
- [x] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: database/schema/12-wasm.sql [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    byte_code               BYTEA           NOT NULL,
+    instantiate_permission  ACCESS_CONFIG   NULL,
+    code_id                 BIGINT          NOT NULL UNIQUE,
+    height                  BIGINT          NOT NULL
+);
+CREATE INDEX wasm_code_height_index ON wasm_code (height);
+
+CREATE TABLE wasm_contract
+(
+    sender                  TEXT            NULL,
```
Comment (MonikaCat) [inline database/schema/12-wasm.sql:29]: Is `sender` value usually `NULL` or should we also add reference to account table?

Comment (huichiaotsou) [reply database/schema/12-wasm.sql:29]: It's null in genesis contracts but should not be empty in general..  but I think we can omit it on the front end though, I see the mintscan does not list it

File: modules/wasm/module.go [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+	"github.com/forbole/juno/v3/modules"
+)
+
+var (
+	_ modules.Module        = &Module{}
+	_ modules.MessageModule = &Module{}
+)
+
+// Module represent x/feegrant module
```
Comment (MonikaCat) [inline modules/wasm/module.go]:
```suggestion
// Module represent x/wasm module
```

File: modules/wasm/source/local/source.go [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+	"github.com/forbole/juno/v3/node/local"
+
+	wasmsource "github.com/forbole/bdjuno/v3/modules/wasm/source"
+)
+
+var (
+	_ wasmsource.Source = &Source{}
+)
+
+// Source implements stakingsource.Source using a local node
```
Comment (MonikaCat) [inline modules/wasm/source/local/source.go]:
```suggestion
// Source implements wasmsource.Source using a local node
```

File: modules/wasm/handle_genesis.go [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+		}
+
+		// Get contract states
+		contractStates, err := m.source.GetContractStates(doc.InitialHeight, contract.ContractAddress)
+		if err != nil {
+			return fmt.Errorf("error while getting genesis contract states: %s", err)
+		}
+
+		contract := types.NewWasmContract(
+			"", contract.ContractInfo.Admin, contract.ContractInfo.CodeID, contract.ContractInfo.Label, nil, nil,
```
Comment (MonikaCat) [inline modules/wasm/handle_genesis.go:107]: Are those values always empty/nil for genesis contracts?

Comment (huichiaotsou) [reply modules/wasm/handle_genesis.go:107]: yep~ only `MsgInstantiate` will include the `sender` value

**Review by MonikaCat — APPROVED** [review]
tested, LGTM

**Review by rmontagnin — APPROVED** [review]
utACK

**Comment by raiftekin** [comment]
```
error while handling MsgExecuteContract: error while storing contracts: error while saving wasm execute contracts: pq: insert or update on table \"wasm_execute_contract\" violates foreign key constraint \"wasm_execute_contract_contract_address_fkey\"
```

**Comment by raiftekin** [comment]
> ```
> error while handling MsgExecuteContract: error while storing contracts: error while saving wasm execute contracts: pq: insert or update on table \"wasm_execute_contract\" violates foreign key constraint \"wasm_execute_contract_contract_address_fkey\"
> ```

image
Also i taking this error.

**Comment by huichiaotsou** [comment]
> ```
> error while handling MsgExecuteContract: error while storing contracts: error while saving wasm execute contracts: pq: insert or update on table \"wasm_execute_contract\" violates foreign key constraint \"wasm_execute_contract_contract_address_fkey\"
> ```

Hi @raifpy this error is because the contract that the MsgExecuteContract wants to execute is not stored in the DB. Parsing from block 1 where that Instantiate Contract messages exist would solve the problem.

**Comment by huichiaotsou** [comment]
> > ```
> > error while handling MsgExecuteContract: error while storing contracts: error while saving wasm execute contracts: pq: insert or update on table \"wasm_execute_contract\" violates foreign key constraint \"wasm_execute_contract_contract_address_fkey\"
> > ```
> 
> image Also i taking this error.

this should have been fixed now, thanks.

**Pull request merged**